File: plugins/codex-usage-sidebar/native/Sources/SidebarCore/IndicatorPlacement.swift
import CoreGraphics
import Foundation

public enum IndicatorPlacementMode: String, CaseIterable, Codable, Sendable {
    case automatic
    case free
    case locked

    public var usesManualFrame: Bool {
        self != .automatic
    }
}

public struct IndicatorManualPlacement: Codable, Equatable, Sendable {
    public let normalizedX: Double
    public let normalizedY: Double

    public init(normalizedX: Double, normalizedY: Double) {
        self.normalizedX = Self.clamped(normalizedX)
        self.normalizedY = Self.clamped(normalizedY)
    }

    public init(frame: CGRect, in visibleFrame: CGRect) {
        let available = Self.availableOriginRect(
            visibleFrame: visibleFrame,
            size: frame.size
        )
        self.init(
            normalizedX: Self.normalized(
                frame.minX,
                lowerBound: available.minX,
                upperBound: available.maxX
            ),
            normalizedY: Self.normalized(
                frame.minY,
                lowerBound: available.minY,
                upperBound: available.maxY
            )
        )
    }

    public func resolvedFrame(in visibleFrame: CGRect, size: CGSize) -> CGRect {
        let available = Self.availableOriginRect(
            visibleFrame: visibleFrame,
            size: size
        )
        return CGRect(
            x: available.minX + available.width * normalizedX,
            y: available.minY + available.height * normalizedY,
            width: size.width,
            height: size.height
        )
    }

    private static func availableOriginRect(
        visibleFrame: CGRect,
        size: CGSize
    ) -> CGRect {
        CGRect(
            x: visibleFrame.minX,
            y: visibleFrame.minY,
            width: max(0, visibleFrame.width - size.width),
            height: max(0, visibleFrame.height - size.height)
        )
    }

    private static func normalized(
        _ value: CGFloat,
        lowerBound: CGFloat,
        upperBound: CGFloat
    ) -> Double {
        let range = upperBound - lowerBound
        guard range > 0 else {
            return 0
        }
        return clamped(Double((value - lowerBound) / range))
    }

    private static func clamped(_ value: Double) -> Double {
        min(1, max(0, value))
    }
}

public struct IndicatorPlacementPreferences: Codable, Equatable, Sendable {
    public var mode: IndicatorPlacementMode
    public private(set) var activeManualDisplayID: String?
    private var placements: [String: IndicatorManualPlacement]

    public init(
        mode: IndicatorPlacementMode = .automatic,
        activeManualDisplayID: String? = nil,
        placements: [String: IndicatorManualPlacement] = [:]
    ) {
        self.mode = mode
        self.activeManualDisplayID = activeManualDisplayID
        self.placements = placements
    }

    public func placement(for displayID: String) -> IndicatorManualPlacement? {
        placements[displayID]
    }

    public mutating func captureManualPlacement(
        frame: CGRect,
        visibleFrame: CGRect,
        displayID: String
    ) {
        placements[displayID] = IndicatorManualPlacement(
            frame: frame,
            in: visibleFrame
        )
        activeManualDisplayID = displayID
    }

    private enum CodingKeys: String, CodingKey {
        case mode
        case activeManualDisplayID
        case placements
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        mode = try container.decodeIfPresent(IndicatorPlacementMode.self, forKey: .mode) ?? .automatic
        activeManualDisplayID = try container.decodeIfPresent(String.self, forKey: .activeManualDisplayID)
        placements = try container.decodeIfPresent([String: IndicatorManualPlacement].self, forKey: .placements) ?? [:]
    }
}

public enum IndicatorPlacementResolver {
    public static func frame(
        preferences: IndicatorPlacementPreferences,
        automaticFrame: CGRect,
        displayID: String,
        visibleFrame: CGRect
    ) -> CGRect {
        guard
            preferences.mode.usesManualFrame,
            let placement = preferences.placement(for: displayID)
        else {
            return automaticFrame
        }
        return placement.resolvedFrame(
            in: visibleFrame,
            size: automaticFrame.size
        )
    }
}

public enum IndicatorPointerInteraction {
    public static func beginsDrag(
        mode: IndicatorPlacementMode,
        origin: CGPoint,
        current: CGPoint,
        threshold: CGFloat = 4
    ) -> Bool {
        guard mode == .free else {
            return false
        }
        return hypot(current.x - origin.x, current.y - origin.y) >= threshold
    }
}

/// Tracks a free-position drag across normal AppKit mouse events. Keeping this
/// state independent of the view avoids a nested event loop that starves
/// layout and drawing while the pointer is moving.
public struct IndicatorDragSession: Equatable, Sendable {
    private var origin: CGPoint?
    public private(set) var isDragging = false

    public init() {}

    public mutating func begin(at point: CGPoint) {
        origin = point
        isDragging = false
    }

    public mutating func update(
        to point: CGPoint,
        mode: IndicatorPlacementMode
    ) -> CGPoint? {
        guard let origin else {
            return nil
        }
        if !isDragging {
            guard IndicatorPointerInteraction.beginsDrag(
                mode: mode,
                origin: origin,
                current: point
            ) else {
                return nil
            }
            isDragging = true
        }
        return CGPoint(
            x: point.x - origin.x,
            y: point.y - origin.y
        )
    }

    @discardableResult
    public mutating func end() -> Bool {
        let didDrag = isDragging
        origin = nil
        isDragging = false
        return didDrag
    }
}

/// Resolves a panel attached to the usage indicator without allowing it to
/// cover the indicator. The preferred placement is below and leading-aligned;
/// the resolver moves it above when the lower screen area cannot contain it.
public enum IndicatorAttachedPanelLayout {
    public static func frame(
        indicatorFrame: CGRect,
        panelSize: CGSize,
        visibleFrame: CGRect,
        screenMargin: CGFloat = 8,
        controlGap: CGFloat = 8
    ) -> CGRect {
        let width = min(
            max(0, panelSize.width),
            max(0, visibleFrame.width - screenMargin * 2)
        )
        let maximumPanelHeight = max(0, visibleFrame.height - screenMargin * 2)
        let desiredHeight = min(max(0, panelSize.height), maximumPanelHeight)
        let minimumX = visibleFrame.minX + screenMargin
        let maximumX = visibleFrame.maxX - width - screenMargin
        let leadingX = indicatorFrame.minX
        let trailingX = indicatorFrame.maxX - width
        let x: CGFloat
        if leadingX >= minimumX, leadingX <= maximumX {
            x = leadingX
        } else if trailingX >= minimumX, trailingX <= maximumX {
            x = trailingX
        } else {
            x = min(maximumX, max(minimumX, leadingX))
        }

        let minimumY = visibleFrame.minY + screenMargin
        let maximumY = visibleFrame.maxY - screenMargin
        let belowAvailable = max(0, indicatorFrame.minY - controlGap - minimumY)
        let aboveAvailable = max(0, maximumY - indicatorFrame.maxY - controlGap)
        let placeBelow = belowAvailable >= desiredHeight || aboveAvailable == 0
        let sideAvailableHeight = placeBelow ? belowAvailable : aboveAvailable
        let height = min(desiredHeight, sideAvailableHeight)
        let y = placeBelow
            ? indicatorFrame.minY - controlGap - height
            : indicatorFrame.maxY + controlGap
        return CGRect(x: x, y: y, width: width, height: height)
    }
}

/// Keeps the opening secondary-click from being interpreted as the first
/// outside click that dismisses the position-mode menu.
public struct PositionModeMenuDismissalState: Equatable, Sendable {
    private var isOpening = false

    public init() {}

    public var shouldDismissForOutsidePointerEvent: Bool {
        !isOpening
    }

    public mutating func beginPresentation() {
        isOpening = true
    }

    public mutating func finishOpeningGesture() {
        isOpening = false
    }

    public mutating func reset() {
        isOpening = false
    }
}

File: plugins/codex-usage-sidebar/windows/src/CodexUsageSidebar.Core/IndicatorPlacement.cs
namespace CodexUsageSidebar.Core;

public enum IndicatorPlacementMode { Automatic, Free, Locked }

public sealed record IndicatorManualPlacement(double NormalizedX, double NormalizedY)
{
    public static IndicatorManualPlacement Capture(RectD frame, RectD workArea) => new(
        Normalize(frame.X - workArea.X, workArea.Width - frame.Width),
        Normalize(frame.Y - workArea.Y, workArea.Height - frame.Height));

    public RectD Resolve(RectD workArea, double width, double height) => new(
        workArea.X + Math.Max(0, workArea.Width - width) * Clamp(NormalizedX),
        workArea.Y + Math.Max(0, workArea.Height - height) * Clamp(NormalizedY), width, height);

    private static double Normalize(double value, double range) => range > 0 ? Clamp(value / range) : 0;
    private static double Clamp(double value) => double.IsFinite(value) ? Math.Clamp(value, 0, 1) : 0;
}

public sealed class IndicatorPlacementPreferences
{
    public IndicatorPlacementMode Mode { get; set; }
    public string? ActiveDisplayId { get; set; }
    public Dictionary<string, IndicatorManualPlacement> Placements { get; set; } = new();

    public void Capture(string displayId, RectD frame, RectD workArea)
    {
        Placements[displayId] = IndicatorManualPlacement.Capture(frame, workArea);
        ActiveDisplayId = displayId;
    }

    public RectD Resolve(string displayId, RectD workArea, RectD automaticFrame) =>
        Mode != IndicatorPlacementMode.Automatic && Placements.TryGetValue(displayId, out var placement)
            ? placement.Resolve(workArea, automaticFrame.Width, automaticFrame.Height)
            : automaticFrame;
}

/// All coordinates are physical screen pixels. A moving window's client
/// coordinates must never be fed back as the next drag origin.
public sealed class IndicatorDragSession
{
    private (double X, double Y, RectD Frame, double Threshold)? origin;
    public bool IsDragging { get; private set; }
    public bool IsActive => origin.HasValue;

    public void Begin(double screenX, double screenY, RectD frame, double dpiScale = 1)
    {
        origin = (screenX, screenY, frame, 4 * (double.IsFinite(dpiScale) && dpiScale > 0 ? dpiScale : 1));
        IsDragging = false;
    }

    public RectD? Update(double screenX, double screenY, IndicatorPlacementMode mode)
    {
        if (origin is not { } start || mode != IndicatorPlacementMode.Free
            || !double.IsFinite(screenX) || !double.IsFinite(screenY)) return null;
        var dx = screenX - start.X;
        var dy = screenY - start.Y;
        if (!IsDragging && Math.Sqrt(dx * dx + dy * dy) < start.Threshold) return null;
        IsDragging = true;
        return start.Frame with { X = start.Frame.X + dx, Y = start.Frame.Y + dy };
    }

    public bool End()
    {
        var didDrag = IsDragging;
        origin = null;
        IsDragging = false;
        return didDrag;
    }
}
